Text Editor using code mirror

barora05 opened the pull request:
## Description
Generic text editor widget using code mirror for prompt. Closes #1134
Highlights as blue if prompt var is valid otherwise as red.

## User Impact
<!-- Describe the impact of this change on the end-users. -->

### Demo
[Demo](https://drive.google.com/file/d/1CeWn1Acq4V6koU9OnN-Qg7o-vRxaOFoA/view?usp=drive_link)
### Docs and Changelog
<!--Link to documentation that has been updated.-->

snopoke commented on assets/javascript/apps/pipeline/nodes/widgets.tsx:
```diff
@@ -1101,3 +1104,219 @@ function BuiltInToolsWidget(props: WidgetParams) {
+
+  useEffect(() => {
+    const updateTheme = () =>
+      setIsDarkMode(document.body.getAttribute("data-theme") === "dark");
+    updateTheme();
+    const observer = new MutationObserver(updateTheme);
+    observer.observe(document.body, { attributes: true });
```
The theme attribute is set on `documentElement`
```suggestion
      setIsDarkMode(document.documentElement.getAttribute("data-theme") === "dark");
    updateTheme();
    const observer = new MutationObserver(updateTheme);
    observer.observe(document.documentElement, { attributes: true });
```

snopoke commented on assets/javascript/apps/pipeline/nodes/widgets.tsx:
```diff
@@ -1101,3 +1104,219 @@ function BuiltInToolsWidget(props: WidgetParams) {
+      >
+        <div className="relative w-full">
+          <textarea
+            className="textarea textarea-bordered resize-none textarea-sm w-full !bg-white !text-black"
```
Textarea colors are controlled by the theme
```suggestion
            className="textarea textarea-bordered resize-none textarea-sm w-full"
```

snopoke commented on assets/javascript/apps/pipeline/nodes/widgets.tsx:
```diff
@@ -1101,3 +1104,219 @@ function BuiltInToolsWidget(props: WidgetParams) {
+  );
+}
+
+const autocompleteVarTheme = EditorView.theme({
```
These need to handle dark mode as well.

snopoke commented on assets/javascript/apps/pipeline/nodes/widgets.tsx:
```diff
@@ -1101,3 +1104,219 @@ function BuiltInToolsWidget(props: WidgetParams) {
+          const to = from + match[0].length;
+
+          const deco = Decoration.mark({
+            class: autocomplete_vars_list.includes(varName)
+              ? "autocomplete-var-valid"
+              : "autocomplete-var-invalid",
```
This should also handle cases where we are using dot notiation e.g. `{session_state.category}` or `{participant_data.schedules[0].name}`.

See https://docs.openchatstudio.com/concepts/prompt_variables/?h=prompt#nested-data-access

snopoke reviewed (commented):
Generally looks good. One issue I found when testing this was that if I scroll or drag in the editor it interacts with the pipeline canvas in the background.

barora05 commented:
> Generally looks good. One issue I found when testing this was that if I scroll or drag in the editor it interacts with the pipeline canvas in the background.

Got it — thanks for pointing that out! I’ll check and fix this issue.

barora05 commented:
@snopoke I have fixed the code and issues related to the dark mode. I have also changed the drag icon to edit icon, but if you still try to drag , there is a weird pipeline drag but the actual nodes and pipeline position is not changed. Here is the updated [demo video ](https://drive.google.com/file/d/1s5dHRnst2NimEEnY31uw1ViZLgcjve_Q/view?usp=drive_link)

snopoke commented on assets/javascript/apps/pipeline/nodes/widgets.tsx:
```diff
@@ -1101,3 +1104,223 @@ function BuiltInToolsWidget(props: WidgetParams) {
+}
+
+const autocompleteVarTheme = (isDarkMode: boolean) =>
+  EditorView.theme({
```
Can we change the font to match the font elsewhere in OCS (by default it's using monospace which is OK for code and JSON but not great for prompts):

I think it's this:
```
.cm-editor .cm-content {
    fontFamily: "ui-sans-serif, system-ui, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji'"
}
```

snopoke replied:
Also, in dark mode let's set the background and text color as follows (lower contrast for easier reading)

```
background-color: oklch(25.33% 0.016 252.42)
color: oklch(97.807% 0.029 256.847)
```

(values taken from DaisyUI `--color-base-100` and `--color-base-content`: https://github.com/saadeghi/daisyui/blob/master/packages/daisyui/src/themes/dark.css)

snopoke reviewed (commented):
Thanks for the updates. It seems to be working better now. A few small styling comments but otherwise looks good.
